File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        l = s.split(' ')
        if len(l) != len(pattern): return False
        mat1 = {}
        mat2 = {}
        for i in range(len(l)):
            if pattern[i] not in mat1 and l[i] not in mat2:
                mat1[pattern[i]] = l[i]
                mat2[l[i]] = pattern[i]
            elif pattern[i] in mat1 and l[i] != mat1[pattern[i]] or l[i] in mat2 and mat2[l[i]] != pattern[i]:
                return False
        return True

File: 290-word-pattern/word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        string cur = "";
        vector<string> words;
        for (auto c : s) {
            if (c == ' '){
                words.push_back(cur);
                cur = "";
            }else{
                cur += c;
            }
        } 
        words.push_back(cur);
        if (words.size() != pattern.size()) return false;
        unordered_map<char, string> hash;
        unordered_map<string, char> hs;
        for (int i = 0; i < pattern.size(); i++){
            auto c = pattern[i];
            if (hash.count(c) == 0 && hs.count(words[i]) == 0){
                hash[c] = words[i];
                hs[words[i]] = c;
            }else if (hash[c] != words[i] || hs[words[i]] != c) {
                cout << hash[c] << " " << words[i];
                return false;
            }
        }
        return true;
    }
};